src/main/java/pl/km/exercise261/RecipeController.java

```diff
@@ -0,0 +1,209 @@
+import static pl.km.exercise261.ApplicationDisplayTemplates.*;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+@Controller
+public class RecipeController {
+    private final RecipeService recipeService;
+    private final RecipeCategoryRepository recipeCategoryRepository;
+    private final IngredientRepository ingredientRepository;
+
+    public RecipeController(RecipeService recipeService,
+                            RecipeCategoryRepository recipeCategoryRepository,
+                            IngredientRepository ingredientRepository) {
+        this.recipeService = recipeService;
+        this.recipeCategoryRepository = recipeCategoryRepository;
+        this.ingredientRepository = ingredientRepository;
+    }
+
+    @ModelAttribute
+    private void addCategoriesToModel(Model model) {
+        List<RecipeCategory> categories = recipeCategoryRepository.findAll();
+        model.addAttribute("categories", categories);
+    }
+
+    @GetMapping("/")
+    public String home(Model model) {
+        List<Recipe> recipes = recipeService.get10Newest();
+        List<String> prepTimes = recipeService.getStringedTimeListFromRecipesList(recipes);
+        model.addAttribute("recipes", recipes);
+        model.addAttribute("prepTimes", prepTimes);
+        model.addAttribute("title", TITLE_HOME);
+        model.addAttribute("pictureName", PICTURE_HOME);
+        return "index";
+    }
+
+    @GetMapping("/popularne")
+    public String showMostLikedRecipes(Model model) {
+        List<Recipe> recipes = recipeService.get5MostPopular();
+        model.addAttribute("recipes", recipes);
+        List<String> prepTimes = recipeService.getStringedTimeListFromRecipesList(recipes);
+        model.addAttribute("prepTimes", prepTimes);
+        model.addAttribute("title", TITLE_MOST_POPULAR);
+        model.addAttribute("pictureName", PICTURE_MOST_POPULAR);
+        return "index";
+    }
+
+    @GetMapping("/kategoria/{id}")
+    public String listRecipesByCategories(@PathVariable Long id, Model model) {
+        Optional<RecipeCategory> recipeCategoryOptional = recipeCategoryRepository.findById(id);
+        if (recipeCategoryOptional.isPresent()) {
+            RecipeCategory category = recipeCategoryOptional.get();
+            model.addAttribute("category", category);
+            List<Recipe> recipes = recipeService.getByCategoryOrderedByName(category);
+            model.addAttribute("recipes", recipes);
+            List<String> prepTimes = recipeService.getStringedTimeListFromRecipesList(recipes);
+            model.addAttribute("prepTimes", prepTimes);
+            String title = category.getName() + " : " + category.getDescription();
+            String pictureName = PICTURE_COMMON_NAME_FOR_CATEGORIES + id;
+            model.addAttribute("title", title);
+            model.addAttribute("pictureName", pictureName);
+            return "index";
+        } else {
+            return "redirect:/";
```

Comment:
jeżeli kategoria nie zostanie odnaleziona, to powinieneś zwrócić kod odpowiedzi 404. W code review pokazałem jak to zrobić w uniwersalny sposób